### PR TITLE
Move to explicit method call for Basic Auth credentials in push client

### LINKED ISSUE #170: Pushgateway basic auth with special characters in password
Looks like we can't to make basic auth with special characters in the password. I've implemented fast monkey-patching
```
module PrometheusClientPush
  attr_reader :basic_auth

  def set_basic_auth(username, password)
    @basic_auth = OpenStruct.new(username: username, password: password)
  end

  private

  def request(req_class, registry = nil)
    req = req_class.new(@uri)
    req.content_type = Prometheus::Client::Formats::Text::CONTENT_TYPE
    req.basic_auth(@uri.user || basic_auth.username, @uri.password || basic_auth.password) if @uri.user || basic_auth
    req.body = Prometheus::Client::Formats::Text.marshal(registry) if registry

    @http.request(req)
  end
end

Prometheus::Client::Push.prepend(PrometheusClientPush)
```

Will be great to see this possibility in a new version of gem, perhaps with better implementation

### PR DESCRIPTION
**Pull request by Sinjo**
While URLs do support passing Basic Auth credentials using the
`http://user:password@example.com/` syntax, the username and password in
that syntax have to follow the usual rules for URL encoding of
characters per RFC 3986
(https://datatracker.ietf.org/doc/html/rfc3986#section-2.1).

Rather than place the burden of correctly performing that encoding on
users of this gem, we decided to have a separate method for supplying
Basic Auth credentials, with no requirement to URL encode the characters
in them.

---

Since the tests around HTTP requests are extremely heavily mocked, I tested this locally with a real pushgateway configured with one user with `@` in the password and another with `:` in it, as well as a more boring password that only used basic latin alphabet characters. All three worked.

Maybe at some point we should have some integration tests, but that feels like its own piece of work.

Fixes #170

### CODE REVIEW

File: lib/prometheus/client/push.rb [hunk 1]
```diff
@@ -39,14 +39,21 @@ def initialize(job:, gateway: DEFAULT_GATEWAY, grouping_key: {}, **kwargs)
...
         @uri = parse("#{@gateway}#{@path}")
+        validate_no_basic_auth!(@uri)
 
         @http = Net::HTTP.new(@uri.host, @uri.port)
         @http.use_ssl = (@uri.scheme == 'https')
         @http.open_timeout = kwargs[:open_timeout] if kwargs[:open_timeout]
         @http.read_timeout = kwargs[:read_timeout] if kwargs[:read_timeout]
       end
 
+      def basic_auth(user, password)
```
Comment (Sinjo) [inline lib/prometheus/client/push.rb:52]: I went with the same method name as `Net::HTTP` as it seems like a reasonable one, though I'd also be open to something like `set_basic_auth` to make it clearer if you'd prefer.

File: lib/prometheus/client/push.rb [hunk 2]
```diff
@@ -126,6 +133,27 @@ def synchronize
         @mutex.synchronize { yield }
       end
 
+      def validate_no_basic_auth!(uri)
+        if uri.user || uri.password
+          raise ArgumentError, <<~EOF
+            Setting Basic Auth credentials in the gateway URL is not supported, please call the `basic_auth` method
```
Comment (nickcampbell18) [inline lib/prometheus/client/push.rb:139]: Driveby 😄 maybe a quick usage example here that a developer could just copy-paste?

**Review by dmagliola — APPROVED** [review]
Beautiful error message

File: lib/prometheus/client/push.rb (hunk 2 above)
Comment (Sinjo) [reply lib/prometheus/client/push.rb]: That's a really good idea. Just added it in!

Also hi Nick 👋🏻
